curl -fsSL get.nextflow.io | bash

./nextflow run -profile docker main.nf --threads 10 --genome GENOME_FILE.fa --reads1 READS1_FILE.fq --reads2 READS2_FILE.fq --outdir OUTPUT_DIR




cd ebola_nextflow/test
./nextflow run -profile docker ../main.nf --threads 20 --genome ebola_ref.fasta --reads1 ebola_mut_reads1.fq --reads2 ebola_mut_reads2.fq --outdir OUTPUT_DIR

main.nf is a file too. If you change directory, you have to change the path or always use the absolute path.



../nextflow run -profile docker ../main.nf --threads 20 --genome ebola_ref.fasta --reads1 ebola_mut_reads1.fq --reads2 ebola_mut_reads2.fq --outdir OUTPUT_DIR

./nextflow -> ../nextflow
main.nf -> ../main.nf



Download and decompress the new version
cd ebola_nextflow
make install
docker build -t ebola_nextflow:latest .
./nextflow run main.nf -with-docker ebola_nextflow:latest
# Ebola - Nextflow pipeline
## Installation
```
unzip ebola_nextflow.zip
cd ebola_nextflow
make install
docker build -t ebola_nextflow:latest .
```
## Run test
```
cd ebola_nextflow
./nextflow run main.nf -with-docker ebola_nextflow:latest
```

Input files used by this test are:
- **Genome:** ebola_nextflow/test/ebola_ref.fasta
- **Reads 1:** ebola_nextflow/test/ebola_mut_reads1.fq
- **Reads 2:** ebola_nextflow/test/ebola_mut_reads2.fq

Files generated by this test are:
- **BAM file:** ebola_nextflow/results/Aligned_sorted_RG_index.marked_split.bam
- **Metrics file:** ebola_nextflow/results/Aligned_sorted_RG_dup_metrics
- **Variants:** ebola_nextflow/results/variants.vcf
- **Indels:** ebola_nextflow/results/indels.vcf

## Run with other parameters
```
cd ebola_nextflow
./nextflow run -profile docker main.nf --genome GENOME_FILE.fa --reads1 READS1_FILE.fq --reads2 READS2_FILE.fq --outdir OUTPUT_DIR
```

## Clean workspace
```
cd ebola_nextflow
make clean
```
